File: token-procedure-archetype/src/main/resources/archetype-resources/src/main/java/__pluginName__/__pluginName__OpenIdUserInfoTokenProcedure.java
package ${package}.${pluginName};

import se.curity.identityserver.sdk.procedure.token.OpenIdUserInfoTokenProcedure;
import se.curity.identityserver.sdk.procedure.token.context.OpenIdUserInfoTokenProcedurePluginContext;
import se.curity.identityserver.sdk.web.ResponseModel;

public final class ${pluginName}OpenIdUserInfoTokenProcedure implements OpenIdUserInfoTokenProcedure
{
    private final ${pluginName}TokenProcedureConfig _configuration;

    public ${pluginName}OpenIdUserInfoTokenProcedure(${pluginName}TokenProcedureConfig configuration)
    {
        _configuration = configuration;
    }

    @Override
    public ResponseModel run(OpenIdUserInfoTokenProcedurePluginContext context)
    {
        var responseData = context.getDefaultResponseData().asMap();

        var name = context.getAccountAttributes().getName();
        if (name != null)
        {
            var formattedName = name.getFormatted();
            if (formattedName != null && !formattedName.isEmpty())
            {
                responseData.put("name", formattedName);
            }
        }

        var presentedTokenData = context.getPresentedToken().getTokenData();

        responseData.put("scope", presentedTokenData.getMandatoryValue("scope", String.class));

        var delegation = context.getPresentedToken().getTokenDelegation();
        if (delegation != null)
        {
            responseData.put("client_id", delegation.getClientId());
        }

        var accountAttributes = context.getAccountAttributes();
        if (accountAttributes != null)
        {
            responseData.put("preferred_username", accountAttributes.getUserName());
        }

        return ResponseModel.mapResponseModel(responseData);
    }
}

File: kotlin-token-procedure-archetype/src/main/resources/archetype-resources/src/main/kotlin/__pluginName__/__pluginName__OpenIdUserInfoTokenProcedure.kt
package ${package}.${pluginName}

import se.curity.identityserver.sdk.procedure.token.OpenIdUserInfoTokenProcedure
import se.curity.identityserver.sdk.procedure.token.context.OpenIdUserInfoTokenProcedurePluginContext
import se.curity.identityserver.sdk.web.ResponseModel

class ${pluginName}OpenIdUserInfoTokenProcedure(private val _configuration: ${pluginName}TokenProcedureConfig): OpenIdUserInfoTokenProcedure
{

    override fun run(context: OpenIdUserInfoTokenProcedurePluginContext): ResponseModel
    {
        val responseData = context.defaultResponseData.asMap()

        context.accountAttributes?.name?.formatted?.let {
            responseData["name"] = it
        }

        val presentedTokenData = context.presentedToken.tokenData
        responseData["scope"] = presentedTokenData.getMandatoryValue("scope", String::class.java)

        context.presentedToken.tokenDelegation?.let {
            responseData["client_id"] = it.clientId
        }

        context.accountAttributes?.let {
            responseData["preferred_username"] = it.userName
        }

        return ResponseModel.mapResponseModel(responseData)
    }
}
